tensorflow/contrib/tensorrt/resources/trt_int8_calibrator.cc

```diff
@@ -0,0 +1,112 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/contrib/tensorrt/resources/trt_int8_calibrator.h"
+
+#include <atomic>
+#include <chrono>
+#include <unordered_map>
+
+#include "tensorflow/core/platform/logging.h"
+
+#if GOOGLE_CUDA
+#if GOOGLE_TENSORRT
+#include "cuda_runtime_api.h"
+
+namespace tensorflow {
+namespace trt {
+
+// set the batch size before constructing the thread to execute engine
+int TRTInt8Calibrator::getBatchSize() const { return batch_size_; }
+
+TRTInt8Calibrator::TRTInt8Calibrator(
+    const std::unordered_map<string, std::pair<void*, size_t>>& dev_buffers,
+    int batch_size, string engine_name)
+    : batch_size_(batch_size),
+      done_(false),
+      dev_buffers_(dev_buffers),
+      calib_running_(false),
+      engine_name_(engine_name) {}
+
+bool TRTInt8Calibrator::setBatch(
+    const std::unordered_map<string, void*>& data) {
+  if (done_) return false;
+  while (calib_running_.load(
+      std::memory_order_acquire)) {  // wait while calibration is running
+    tensorflow::mutex_lock l(cond_mtx_);
+    cond_.wait_for(l, std::chrono::milliseconds(50));
```

Comment:
Will this become obsolete in future PRs? If not, I think there are a few improvements we can make to make the logic clear, less error-prone, and easier to maintain in the future:

1. we only need a wait() in setBatch
2. the mutex_lock should be outside of the loop, as otherwise the logic is identical (as safe as) without the mutex (see below)
3. done_ needs to be protected by the mutex
4. we should not miss the first batch, even though this may be statistically ok, but it's functionally incorrect. I think we can use this condition_variable to fix that.

My another argument is that using atomic words (aka lock-free) instead of mutex is error-prone and make the code harder to maintain. I think the current approach is identical to the code snippet below which doesn't use mutex/condition-var:
```
constructor() {
  calib_running_ = false;
  done_ = false;
}
setBatch() {
  if (done_) return false;
  while (calib_running_.load(acquire)) {
    sleep(50ms);
    if (done_) return false;
  }
  // set the batch
  calib_running_.store(true, release);
}
getBatch() {
  calib_running_.store(false, release);
  while (!calib_running_.load(acquire)) {
    sleep(50ms);
    if (done_) return false;
  }
  if (done_) return false;
  // get the batch
}
setDone() {
  done_ = true;
}
```
I think we can fix all the issues mentioned above by doing:
```
constructor() {
  // All these variables are protected by the same mutex.
  batch_is_set_ = false;
  calib_running_ = false;
  done_ = false;
}
setBatch() {
  mutex_lock l(cond_mtx_);
  while ((calib_running_ || batch_is_set_) && !done_) {
    cond_.wait(l);
  }
  if (done_) return false;
  CHECK(!calib_running_ && !batch_is_set_);
  // set the batch
  batch_is_set_ = true;
  cond_.notify_all();
}
getBatch() {
  mutex_lock l(cond_mtx_);
  calib_running_ = false;
  cond_.notify_all();  // Set the batch if !batch_is_set_
  while (!batch_is_set_ && !done_) {
    cond_.wait(l);
  }
  if (done_) return false;
  CHECK(!calib_running_ && batch_is_set_);
  // get the batch
  batch_is_set_ = false;
  calib_running_ = true;
}
setDone() {
  mutex_lock l(cond_mtx_);
  done_ = true;
  cond_.notify_all();
}
```
Please take a look @samikama. Thanks :)